### ISSUE TITLE
Muso production users have multiple task documents in non-terminal state with the same emission id

### ISSUE THREAD

**Issue by kennsippell** [medic/cht-core#6562]
**Describe the bug**
Muso users are reporting that they are unable to complete tasks after upgrading to v3.8.1. This should be fixed by #6389 available in v3.9 and the Muso project team is pursuing an upgrade to address the symptoms. The problematic state is that there are multiple task documents that are 1) in a non-terminal state at the same time, and 2) have the same emission id.  When users get into this state (even in v3.9), it results in unnecessary task documents that bloat the system (disk, network, analytics). 

We know of one user scenario which leads to this, and that happens as a race condition when the same user has multiple sessions open at the same time (#6389 for details). However, Muso users are (by report) triggering this condition within a single session. It is unknown how that is happening. 

**To Reproduce**
It is unclear how to reproduce the behaviour, but here is an example of it happening for a production user:

[Task Document 1](https://muso-mali.app.medicmobile.org/_utils/#database/medic/task~org.couchdb.user%3Afantatangara239~9c8e55bb-a1b1-45d4-9f25-e85cc0350f61~0~anc_followup_yields_self~1595056258855)
```
{
  "_id": "task~org.couchdb.user:fantatangara239~9c8e55bb-a1b1-45d4-9f25-e85cc0350f61~0~anc_followup_yields_self~1595056258855",
  "_rev": "1-478855e192e2de00b849c4dd3b88a4a5",
  "type": "task",
  "authoredOn": 1595056258855,
  "stateHistory": [
    {
      "state": "Draft",
      "timestamp": 1595056258855
    }
  ],
```

[Task Document 2](https://muso-mali.app.medicmobile.org/_utils/#database/medic/task~org.couchdb.user%3Afantatangara239~9c8e55bb-a1b1-45d4-9f25-e85cc0350f61~0~anc_followup_yields_self~1595569713110)

```
{
  "_id": "task~org.couchdb.user:fantatangara239~9c8e55bb-a1b1-45d4-9f25-e85cc0350f61~0~anc_followup_yields_self~1595569713110",
  "_rev": "1-fa9057ad8d8bf5fef047cd916172e396",
  "type": "task",
  "authoredOn": 1595569713110,
  "stateHistory": [
    {
      "state": "Draft",
      "timestamp": 1595569713110
    }
  ],
```

More details in https://github.com/medic/config-muso/issues/55.
Speculatively, this may be related to navigating between tabs while the task documents are loading.

**Logs**
If applicable, include the server or browser logs.

**Screenshots**

**Environment**
- Instance: https://muso-mali.app.medicmobile.org/
- Browser: Medic-Android v0.4.?
- Client platform: Android
- App: webapp
- Version: 3.8.1

**Comment by dianabarsan** [comment]
I can replicate this reliably on current master.

Replication steps:
1. Add a task that will create 20k events for your user's contact:
```
{
    name: 'two',
    icon: 'icon-follow-up',
    title: 'Many events',
    appliesTo: 'contacts',
    appliesToType: ['person'],
    appliesIf: function (contact) {
      return user._id === contact.contact._id;
    },
    resolvedIf: function () {
      return false;
    },
    actions: [
      {
        type: 'report',
        form: 'self-assessment',
      }
    ],
    events: Array.from({ length: 20000 }).map((el, idx) => {
      const now = new Date('2020-07-01').getTime();
      const days = idx * 24 * 60 * 60 * 1000;
      const event = {
        start: 0,
        end: 0,
        id: `two_${idx}`,
        dueDate: () => new Date(now + days),
      };
      return event;
    }),
  }
```
2. Login as an offline user on a slow enough device (I slowed down my desktop browser by 4x in Performance settings, for example). 
3. Immediately after being logged in navigate to the Tasks tab. immediately navigate to another tab (any tab) then immediately back to Tasks. Your browser may be very slow to respond but ... you know .. insist :D
4. Let the browser do it's thing. It'll take a while. You can check the console for messages like: 
```
Committing xxxxx task document updates
```
If you see two of these logs, it's jackpot and you triggered the bug. 
5. Sync up. This will also take a while.
6. Once synced, check docs on the server by querying `_all_docs` with start_key + end_key like:
```
start_key=`task~org.couchdb.user:<your user's name>~<your user's contact id>~two`
end_key=`task~org.couchdb.user:<your user's name>~<your user's contact id>~two\ufff0`
```

Check the number of returned results. If there are more than 20k, you probably have duplicates. 
You can identify duplicates by their event_id : `task~org.couchdb.user:<username>~<contact_id>~two_10001` (where `two_10001` should be unique!) 

I've labeled this as Critical, as it's affecting a live instance, but lowered priority to avoid spamming channels. 

This is probably happening because we start task recalculation threads every time we navigate to Tasks, regardless if tasks are already calculating. 
I think the solution is to keep track of task calculation threads and never start two for the same contacts in parallel. 

@garethbowen please advise on scheduling. Thank you.

**Comment by garethbowen** [comment]
I think we should start dev right away so I'm scheduling this for 3.10.0. Once that's complete we can consider backporting to 3.9.x if appropriate.

**Comment by garethbowen** [comment]
Ready for AT in `6562-single-task-recalculation-thread`

**Comment by ngaruko** [comment]
Thanks @dianabarsan for the details above. Was able to replicate it on `master`
![image](https://user-images.githubusercontent.com/6979995/89361716-120d2c80-d720-11ea-9661-992048404fbf.png)
 and found some duplicate
```
  {
            "id": "task~org.couchdb.user:andra~2e0ceb06-ced2-5a63-bca0-0283a5aab0e8~two_10000~two~1596590122300",
            "key": "task~org.couchdb.user:andra~2e0ceb06-ced2-5a63-bca0-0283a5aab0e8~two_10000~two~1596590122300",
            "value": {
                "rev": "1-12f735fac597f92a5be1d4743ed4aaf8"
            }
        },
        {
            "id": "task~org.couchdb.user:andra~2e0ceb06-ced2-5a63-bca0-0283a5aab0e8~two_10000~two~1596590122304",
            "key": "task~org.couchdb.user:andra~2e0ceb06-ced2-5a63-bca0-0283a5aab0e8~two_10000~two~1596590122304",
            "value": {
                "rev": "2-c24fd1f423e67612629c9bd765fcd888"
            }
        },
```
I have pinged @senseibara to give it a second look before we close it off

**Comment by senseibara** [comment]
LGTM, i'm not seeing any duplicates, though i noticed during the recalculation that the app was "freezed" for a while, other tabs was not responding. Unsure what happened but wanted to flag that.

**Comment by ngaruko** [comment]
@senseibara I realised that too, but thought it was because of the number of docs (20K plus). We will keep an eye on this during Release Testing.

**Comment by dianabarsan** [comment]
> other tabs was not responding

All tabs require views to be fully indexed in order to get data to be displayed. While you're continually inserting docs in the db, view indexes will try to keep up and keep indexing. 
If you try loading a tab that does not require views, for example the About page, does it load? (expect some slower loading, anyway).

**Comment by senseibara** [comment]
What is expected is that when you click another page you move over there but see no data until loading is finished. 
What I meant was that the page is not being displayed at all, no reaction after click, it is just freezed at the task page, the menu button was not responding either. Is that what you had too @ngaruko  ?

**Comment by dianabarsan** [comment]
I had that as well:
> Your browser may be very slow to respond but ... you know .. insist :D
I've attributed it to the massive calculations the browser has to do in a short timeframe that blocks the main thread. 
Definitely something to look into for "regular" task generation (not 20k tasks at once).

**Comment by dianabarsan** [comment]
Merged into `master`.
